### ISSUE TITLE
Latency continues to rise when adding many charts

### ISSUE THREAD

**Issue by hobti01** [helm/chartmuseum#456]
Using chartmuseum as part of Harbor, we use replication to copy many (thousands) of charts from one instance to another.
During the replication, the chart existence in the destination is tested with a GET before copying takes place.

We see the latency on GET requests rising steadily.  Here is a sampling (every 250th line) of logs to show the latency increase over time.
```
{"L":"INFO","T":"2021-05-10T07:32:07.843Z","M":"[770] Request served","path":"/api/myrepo/charts/mychart/10.2.0","comment":"","latency":"50.965665482s","clientIP":"10.42.25.36","method":"GET","statusCode":200,"reqID":"ce0d34a1-196f-4afc-8d62-4c96922406dc"}
{"L":"INFO","T":"2021-05-10T07:33:52.514Z","M":"[1254] Request served","path":"/api/myrepo/charts/mychart/0.0.0-1af3b533","comment":"","latency":"1m27.551545015s","clientIP":"10.42.25.36","method":"GET","statusCode":200,"reqID":"5d137558-5787-4000-80d6-ccbfc5233c52"}
{"L":"INFO","T":"2021-05-10T07:35:44.657Z","M":"[1795] Request served","path":"/api/myrepo/charts/mychart/0.0.0-5f691ff5","comment":"","latency":"2m10.984774113s","clientIP":"10.42.25.36","method":"GET","statusCode":200,"reqID":"c87631be-b394-4cd2-be05-6c10b7968dd3"}
{"L":"INFO","T":"2021-05-10T07:37:36.581Z","M":"[2333] Request served","path":"/api/myrepo/charts/mychart/2.11.0","comment":"","latency":"2m52.431529688s","clientIP":"10.42.21.61","method":"GET","statusCode":200,"reqID":"0d82e014-65e9-4fe0-9b11-035f84baccac"}
{"L":"INFO","T":"2021-05-10T07:39:29.995Z","M":"[2862] Request served","path":"/api/myrepo/charts/mychart/1.12.0","comment":"","latency":"3m35.025049754s","clientIP":"10.42.21.61","method":"GET","statusCode":200,"reqID":"2bc4a035-cefa-4fc1-864f-1b4414026076"}
{"L":"INFO","T":"2021-05-10T07:41:23.182Z","M":"[3376] Request served","path":"/api/myrepo/charts/mychart/0.0.0-55b702f4","comment":"","latency":"4m22.007597889s","clientIP":"10.42.21.61","method":"GET","statusCode":200,"reqID":"be3dd229-aa27-4d7d-8a35-89c945c341c8"}
{"L":"INFO","T":"2021-05-10T07:43:20.183Z","M":"[3906] Request served","path":"/api/myrepo/charts/mychart/12.7.0","comment":"","latency":"5m21.019376238s","clientIP":"10.42.25.36","method":"GET","statusCode":200,"reqID":"b868e793-7d1e-4498-ad89-3ad88b5e5ab2"}
{"L":"INFO","T":"2021-05-10T07:45:10.301Z","M":"[4401] Request served","path":"/api/myrepo/charts/mychart/9.7.1","comment":"","latency":"6m19.640680281s","clientIP":"10.42.21.61","method":"GET","statusCode":200,"reqID":"81ca37c4-e4dc-4f08-8541-54a8ee3b7d24"}
{"L":"INFO","T":"2021-05-10T07:47:02.389Z","M":"[4925] Request served","path":"/api/myrepo/charts/mychart/7.0.1","comment":"","latency":"7m20.081200197s","clientIP":"10.42.25.36","method":"GET","statusCode":200,"reqID":"1dd354cd-77bf-4c77-a330-78e9e9af37f1"}
{"L":"INFO","T":"2021-05-10T07:48:56.750Z","M":"[5439] Request served","path":"/api/myrepo/charts/mychart/6.0.1","comment":"","latency":"8m20.471966761s","clientIP":"10.42.21.61","method":"GET","statusCode":200,"reqID":"9b1762fd-7607-4581-850d-083cdba79a3f"}
{"L":"INFO","T":"2021-05-10T07:50:51.652Z","M":"[5965] Request served","path":"/api/myrepo/charts/mychart/0.0.0-9fe92d91","comment":"","latency":"9m18.307049618s","clientIP":"10.42.21.61","method":"GET","statusCode":200,"reqID":"f0577a3b-ebfb-4995-97f9-04890d6b84e8"}
{"L":"INFO","T":"2021-05-10T07:52:43.481Z","M":"[6487] Request served","path":"/api/myrepo/charts/mychart/1.7.0","comment":"","latency":"10m13.92548115s","clientIP":"10.42.25.36","method":"GET","statusCode":200,"reqID":"f6a8f9e5-5d58-4e28-b744-8f1a28512620"}
```

My understanding is this issue was resolved in https://github.com/helm/chartmuseum/pull/19

But we continue to see this issue in ChartMuseum `0.12.0` using AWS S3 as storage and redis as an index cache.  This happens on each ChartMuseum instance that is using the S3 storage.
```
ALLOW_OVERWRITE=true
CACHE: redis
DEPTH: "1"
DISABLE_STATEFILES: "false"
INDEX_LIMIT: "0"
STORAGE: amazon
```

Upgrading to ChartMuseum `0.13.1` is unfortunately not an option due to duplicate tag issue https://github.com/helm/chartmuseum/issues/450

**Comment by scbizu** [comment]
This should be fixed in #339 but released in v0.13.0 . The duplicate tag issue fix will be released in v0.13.2 next week since I have another several optimization blocked , give me some more time plz XD

**Comment by hobti01** [comment]
BTW, I've since been testing with
```
DISABLE_STATEFILES: "true"
```
Which reduced the S3 traffic from continuous > 350MB to basically zero.  Because we're using redis as an index cache the overhead of the `index-cache.yaml` on S3 seemed unnecessary.  I also observed rate limiting to the S3 api with the frequent access and update of the index cache in S3.

Changing the `INDEX_LIMIT` changes memory use but does not seem to affect throughput enough to have an effect.

It would be a great metric to see the size of the queue waiting for storage indexing, but if the overall issue is solved in v0.13.x then the metrics are only useful if combined with tuning recommendations to keep the queue size "reasonable".

Thanks for all your efforts.
